src/Connection/DefaultConnectionPool.php

```diff
@@ -2,40 +2,107 @@
 
 namespace Amp\Http\Client\Connection;
 
+use Amp\ByteStream\StreamException;
+use Amp\CancellationToken;
+use Amp\CancelledException;
+use Amp\Http\Client\Internal\CombinedCancellationToken;
 use Amp\Http\Client\Request;
-use Amp\Socket\Socket;
+use Amp\Http\Client\SocketException;
+use Amp\Http\Client\TimeoutException;
+use Amp\Promise;
+use Amp\Socket\ClientTlsContext;
+use Amp\Socket\ConnectContext;
+use Amp\Socket\Connector;
+use Amp\Socket;
+use Amp\Socket\EncryptableSocket;
+use Amp\TimeoutCancellationToken;
+use function Amp\call;
 
 final class DefaultConnectionPool implements ConnectionPool
 {
+    private const APPLICATION_LAYER_PROTOCOLS = ['http/1.1'];
+
+    /** @var Connector */
+    private $connector;
+
+    /** @var Connection[][] */
     private $connections = [];
 
-    public function createConnection(Socket $socket): Connection
+    public function __construct(?Connector $connector = null)
     {
-        $connection = new Http1Connection($socket);
-        $this->connections[$socket->getRemoteAddress()->toString()] = $connection;
-        return $connection;
+        $this->connector = $connector ?? Socket\connector();
     }
 
-    public function getConnection(Request $request): ?Connection
+    public function getConnection(Request $request, CancellationToken $cancellation): Promise
     {
-        $uri = $request->getUri();
+        return call(function () use ($request, $cancellation) {
+            $isHttps = $request->getUri()->getScheme() === 'https';
+            $defaultPort = $isHttps ? 443 : 80;
 
-        $address = $uri->getHost();
-        $port = $uri->getPort();
+            $authority = $request->getUri()->getHost() . ':' . ($request->getUri()->getPort() ?: $defaultPort);
 
-        if ($port !== null) {
-            $address .= ':' . $port;
-        }
+            if (isset($this->connections[$authority])) {
```

Comment:
We should also include the scheme in the key here, there are very rare cases where requests with `http` and `https` are supported on the same port.